crates/services/sync/src/import/cache.rs

```diff
@@ -0,0 +1,541 @@
+
+    pub fn get_chunks(
+        &self,
+        range: RangeInclusive<u32>,
+        max_chunk_size: usize,
+    ) -> futures::stream::Iter<std::vec::IntoIter<CachedDataBatch>> {
+        let end = (*range.end()).saturating_add(1);
+        let chunk_size_u32 = u32::try_from(max_chunk_size)
+            .expect("The size of the chunk can't exceed `u32`");
+
+        let cache_iter = self.collect_cache_data(range.clone());
+        let mut current_height = *range.start();
+        let mut chunks = Vec::new();
+        let mut current_chunk = CachedDataBatch::None(0..0);
+
+        for (height, data) in cache_iter {
+            // We have a range missing in our cache.
+            // Push the existing chunk and push chunks of missing data
+            if height != current_height {
+                Self::push_current_chunk(&mut chunks, current_chunk, max_chunk_size);
+                current_chunk = CachedDataBatch::None(0..0);
+                Self::push_missing_chunks(
+                    &mut chunks,
+                    current_height,
+                    height,
+                    max_chunk_size,
+                    chunk_size_u32,
+                    end,
+                );
+                current_height = height;
+            }
+            // Either accumulate in the same chunk or transition if the data is not the same as the current chunk
+            current_chunk = Self::handle_current_chunk(
+                current_chunk,
+                data,
+                height,
+                &mut chunks,
+                max_chunk_size,
+            );
+            current_height = height.saturating_add(1);
+        }
+
+        Self::push_current_chunk(&mut chunks, current_chunk, max_chunk_size);
+        if current_height <= end {
+            Self::push_missing_chunks(
+                &mut chunks,
+                current_height,
+                end,
+                max_chunk_size,
+                chunk_size_u32,
+                end,
+            );
+        }
+        futures::stream::iter(chunks)
+    }
+
+    // Split the chunk in chunks of `max_chunk_size` and push them in `chunks` vector.
+    fn push_current_chunk(
+        chunks: &mut Vec<CachedDataBatch>,
+        chunk: CachedDataBatch,
+        max_chunk_size: usize,
+    ) {
+        if chunk.is_range_empty() {
+            return;
+        }
+        // Adds a vec of chunks that respect `max_chunk_size`
+        match chunk {
+            CachedDataBatch::Blocks(blocks) => chunks.extend(
+                blocks
+                    .range
+                    .zip(blocks.results)
+                    .chunks(max_chunk_size)
+                    .into_iter()
+                    .map(|chunk| {
+                        let (range, blocks): (Vec<u32>, Vec<SealedBlock>) = chunk.unzip();
+                        CachedDataBatch::Blocks(Batch::new(
+                            None,
+                            range[0]..range[range.len() - 1].saturating_add(1),
```

Comment:
As clippy noted, you'll need to fix the inner subtraction to be saturating, and use `.get()` instead of array slicing here `[]` to avoid potential unexpected panics here.

Reply:
Done.